BFDD : BFD session stays until the last static route over nexhop

varuntumbe opened the pull request:
**What is the problem ?**

When we have more then one static route over the same nexthop with BFD, session goes down and gets cleared if we remove one of the static route. 

**What is the Root Cause ?**

1.  When we create static route over a nexthop with BFD, **bfdd_dest_register** routine will be called where for the first flow, we create new pcn object ( pcn_new ), we increment the refcount and notify the client via **ptm_bfd_notify**

But for the subsequent addition of static route over the same nh, means **pcn_lookup** would give the same existing pcn skipping the increment of refcount.

so essentially, refcount remains at value 1, even if you add any number of static route over the nh.

2.  When we remove one of the static, **bfdd_dest_deregister** will called where we do **pcn_free** ( where we decrement the refcount ) and session delete ( _ptm_bfd_session_del )



**What is the fix ?**

Fix has 2 parts 

1.  Incrementing the refcount properly ( early before pcn_lookup) for every static route addition. This makes sures that refcount gets updated properly for every addition of static route

2.  Now when we start removing the static routes one by one, we decrement the refcount ( in bfdd_dest_deregister ) and will call the pcn_free only when refcount becomes 0 ( essentially deleting the session only when there are no static routes points to nh )

Closes https://github.com/FRRouting/frr/issues/19014

ton31337 reviewed (changes requested):
Please update the commit also (instead of TBD).

varuntumbe commented on tests/topotests/bfd_static_topo/rt1/frr.conf:
```diff
@@ -0,0 +1,29 @@
+log file /tmp/topotests/bfd_ospf_topo1.test_bfd_ospf_topo1/rt1/ospf6d.log
+log file /tmp/topotests/bfd_ospf_topo1.test_bfd_ospf_topo1/rt1/staticd.log
+log file /tmp/topotests/bfd_ospf_topo1.test_bfd_ospf_topo1/rt1/bfdd.log
+no service integrated-vtysh-config
```
Removed

varuntumbe commented on tests/topotests/bfd_static_topo/rt2/frr.conf:
```diff
@@ -0,0 +1,29 @@
+exit
+!
+interface lo
+ ip address 2.2.2.2/32
```
Removed

varuntumbe commented on tests/topotests/bfd_static_topo/test_bfd_static_topo1.py:
```diff
@@ -0,0 +1,169 @@
+# SPDX-License-Identifier: ISC
+
+#
+# test_bfd_static_vrf.py
```
Changed the name

varuntumbe commented on bfdd/ptm_adapter.c:
```diff
@@ -936,6 +937,8 @@ static struct ptm_client_notification *pcn_new(struct ptm_client *pc,
 					       struct bfd_session *bs)
 {
 	struct ptm_client_notification *pcn;
+	/* Lets increment the refcount in the beginning itself regardless of its a new pcn or old*/
```
Earlier flow :
bfdd_dest_register
pcn_new
   ==> here we try to find whether there is already a existing pcn for the particular bsd session, if found, we used to return without updating the ref_count ( refcount updation was later pcn_lookup ) which mess up the refcount logic.

===> if we add multiple static route over the same bfd peer nh, for the subsequent addition, pcn_lookup will give the existing pcn which result in NOT updating the refcount

varuntumbe replied:
Now refcount  incrementation has been added before the pcn_lookup which makes the refcount increment properly

varuntumbe commented:
======> cmd : show bfd peers json

[
  {
    "multihop":false,
    "peer":"2001:db8:1::1",
    "local":"2001:db8:1::2",
    "vrf":"default",
    "interface":"r2-eth0",
    "id":2411218638,
    "remote-id":38497228,
    "passive-mode":false,
    "log-session-changes":false,
    "status":"up",
    "uptime":4,
    "diagnostic":"ok",
    "remote-diagnostic":"ok",
    "type":"dynamic",
    "receive-interval":600,
    "transmit-interval":600,
    "echo-receive-interval":50,
    "echo-transmit-interval":0,
    "detect-multiplier":3,
    "remote-receive-interval":600,
    "remote-transmit-interval":600,
    "remote-echo-receive-interval":50,
    "remote-detect-multiplier":3,
    "rtt-min":0,
    "rtt-avg":0,
    "rtt-max":0
  },
  {
    "multihop":false,
    "peer":"2001:db8:2::1",
    "vrf":"default",
    "id":3088557958,
    "remote-id":2847372379,
    "passive-mode":false,
    "log-session-changes":false,
    "status":"up",
    "uptime":3,
    "diagnostic":"ok",
    "remote-diagnostic":"ok",
    "type":"dynamic",
    "receive-interval":2000,
    "transmit-interval":2000,
    "echo-receive-interval":50,
    "echo-transmit-interval":0,
    "detect-multiplier":3,
    "remote-receive-interval":2000,
    "remote-transmit-interval":2000,
    "remote-echo-receive-interval":50,
    "remote-detect-multiplier":3,
    "rtt-min":0,
    "rtt-avg":0,
    "rtt-max":0
  }
]

================> from the router



[
    {
        "detect-multiplier": 3,
        "diagnostic": "ok",
        "echo-receive-interval": 50,
        "echo-transmit-interval": 0,
        "id": "*",
        "interface": "r2-eth0",
        "local": "2001:db8:1::2",
        "multihop": false,
        "passive-mode": false,
        "peer": "2001:db8:1::1",
        "receive-interval": 600,
        "remote-detect-multiplier": 3,
        "remote-diagnostic": "ok",
        "remote-echo-receive-interval": 50,
        "remote-id": "*",
        "remote-receive-interval": 600,
        "remote-transmit-interval": 600,
        "status": "up",
        "uptime": "*",
        "transmit-interval": 600
    },
    {
        "detect-multiplier": 3,
        "diagnostic": "ok",
        "echo-receive-interval": 50,
        "echo-transmit-interval": 0,
        "id": "*",
        "interface": "r2-eth1",
        "local": "2001:db8:2::2",
        "multihop": false,
        "passive-mode": false,
        "peer": "2001:db8:2::1",
        "receive-interval": 2000,
        "remote-detect-multiplier": 3,
        "remote-diagnostic": "ok",
        "remote-echo-receive-interval": 50,
        "remote-id": "*",
        "remote-receive-interval": 2000,
        "remote-transmit-interval": 2000,
        "status": "up",
        "uptime": "*",
        "transmit-interval": 2000
    }
]


===============> expected o/p from the testcase file


one of the topotest failing consistantly ( bfd_topo3 ) where expected does not match the received o/p from the router.

When I checked, its failing because its not able to find "local"  and/or "interface" fields ( which are in expected files but not in router gen o/p )

I dont see how my changes affected this failure. Any idea  @ton31337 ?

Thanks,
Varun

varuntumbe commented:
> @varuntumbe are you still working on this?

Hi @ton31337 ,

This solution is fundamentally wrong ( as I found out when one of the bfd topo test failed ). This needs a rework. I tried reworking this couple of months back but was not able to code it up to a proper solution. I am not working on this now. Apologies.

thanks

Commit:
topotests: Adding a test to determine the BFD behavior with static routes over same nh

    Adding the topotest which verifies whether BFD session stays up as long as
    there is one static route pointing to nexthop bfd peer

    Signed-off-by: Varun Hegde <varuntumbe1@gmail.com>

Commit:
bfdd: Making the BFD session stay up until the last static route

    This Fix makes the BFD session to stay up until the
    last static route going over the gateway ip which is tracked by bfd.

    Signed-off-by: Varun Hegde <varuntumbe1@gmail.com>

RodrigoMNardi commented:
ci:rerun
